dashboard/final-example/app/lib/actions.ts

```diff
@@ -124,10 +125,15 @@ export async function authenticate(
   formData: FormData,
 ) {
   try {
-    await signIn('credentials', Object.fromEntries(formData));
+    await signIn('credentials', formData);
   } catch (error) {
-    if ((error as Error).message.includes('CredentialsSignin')) {
-      return 'CredentialsSignin';
+    if (error instanceof AuthError) {
+      switch (error.type) {
+        case 'CredentialsSignin':
+          return 'Invalid credentials.';
+        default:
+          return 'Something went wrong.';
+      }
```

Comment:
Should there be a return type outside of this condition (or in an `else`) to catch when `error instanceof AuthError` returns false?

Reply:
Those errors are thrown further down. It's not the task of the `authenticate` method. There are special throws as well that powers `redirect()` (used internally by `signIn`) which should not be handled and let Next.js/React handle them with their error boundaries.

If the error is an `AuthError`, we already know how to handle it.